File: runtime-agent/internal/domain/run_configuration.go
package domain

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"path/filepath"
	"regexp"
	"strings"
	"unicode/utf8"
)

const RunConfigurationVersion = 1

var (
	ErrRunConfigurationsMissing = errors.New("run configurations file not found")
	ErrRunConfigurationsCorrupt = errors.New("run configurations file is corrupt")
	ErrRunConfigurationNotFound = errors.New("run configuration not found")
	ErrRunConfigurationConflict = errors.New("run configuration conflict")
	ErrInvalidRunConfiguration  = errors.New("invalid run configuration")
	stableRunConfigurationID    = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$`)
	contextPathPattern          = regexp.MustCompile(`^/[A-Za-z0-9._-]*$`)
	environmentNamePattern      = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)
	environmentReferencePattern = regexp.MustCompile(`^\$\{env:[A-Za-z_][A-Za-z0-9_]*\}$`)
	sensitiveEnvironmentName    = regexp.MustCompile(`(?i)(PASSWORD|PASSWD|SECRET|TOKEN|API_KEY|PRIVATE_KEY)`)
)

type RunConfigurationDocument struct {
	Version                 int                      `json:"version"`
	Configurations          []TomcatRunConfiguration `json:"configurations"`
	SelectedConfigurationID *string                  `json:"selectedConfigurationId"`
}

type TomcatRunConfiguration struct {
	ID                string                 `json:"id"`
	Name              string                 `json:"name"`
	Type              string                 `json:"type"`
	ProjectID         string                 `json:"projectId"`
	Mode              string                 `json:"mode"`
	Suspend           bool                   `json:"suspend"`
	JDKRef            string                 `json:"jdkRef"`
	Build             RunConfigurationBuild  `json:"build"`
	Server            RunConfigurationServer `json:"server"`
	Deploy            RunConfigurationDeploy `json:"deploy"`
	Env               map[string]string      `json:"env"`
	VMOptions         []string               `json:"vmOptions"`
	BeforeLaunchTasks []string               `json:"beforeLaunchTasks"`
}

type RunConfigurationBuild struct {
	Type    string `json:"type"`
	Target  string `json:"target,omitempty"`
	Command string `json:"command,omitempty"`
	Clean   bool   `json:"clean"`
}

type RunConfigurationServer struct {
	ID          string `json:"id"`
	HTTPPort    int    `json:"httpPort"`
	DebugPort   int    `json:"debugPort"`
	ContextPath string `json:"contextPath"`
}

type RunConfigurationDeploy struct {
	Mode     string `json:"mode"`
	Artifact string `json:"artifact"`
}

func invalidRunConfiguration(path, message string) error {
	return fmt.Errorf("%w: %s %s", ErrInvalidRunConfiguration, path, message)
}

func validateStableID(path, value string) error {
	if !stableRunConfigurationID.MatchString(value) {
		return invalidRunConfiguration(path, "must be a stable id (1-128 ASCII letters, digits, _, . or -)")
	}
	return nil
}

func ValidateRunConfigurationID(value string) error {
	return validateStableID("id", value)
}

func validateNonBlank(path, value string, max int) error {
	if value == "" || strings.TrimSpace(value) != value || utf8.RuneCountInString(value) > max {
		return invalidRunConfiguration(path, fmt.Sprintf("must be non-blank without surrounding whitespace and at most %d characters", max))
	}
	return nil
}

func validateArtifact(value string) error {
	if err := validateNonBlank("deploy.artifact", value, 1024); err != nil {
		return err
	}
	if filepath.IsAbs(value) || strings.HasPrefix(value, "/") || strings.HasPrefix(value, `\`) || strings.Contains(value, `\`) {
		return invalidRunConfiguration("deploy.artifact", "must be a forward-slash project-root-relative path")
	}
	if strings.Contains(value, ":") {
		return invalidRunConfiguration("deploy.artifact", "must not contain a colon")
	}
	for _, segment := range strings.Split(value, "/") {
		if segment == ".." {
			return invalidRunConfiguration("deploy.artifact", "must not contain .. traversal segments")
		}
	}
	return nil
}

func (configuration TomcatRunConfiguration) Validate() error {
	if err := validateStableID("id", configuration.ID); err != nil {
		return err
	}
	if err := validateNonBlank("name", configuration.Name, 200); err != nil {
		return err
	}
	if configuration.Type != "tomcat6" {
		return invalidRunConfiguration("type", "must equal tomcat6")
	}
	if err := validateStableID("projectId", configuration.ProjectID); err != nil {
		return err
	}
	if configuration.Mode != "run" && configuration.Mode != "debug" {
		return invalidRunConfiguration("mode", "must be run or debug")
	}
	if configuration.Mode == "run" && configuration.Suspend {
		return invalidRunConfiguration("suspend", "must be false in run mode")
	}
	if err := validateStableID("jdkRef", configuration.JDKRef); err != nil {
		return err
	}

	switch configuration.Build.Type {
	case "ant":
		if err := validateNonBlank("build.target", configuration.Build.Target, 200); err != nil {
			return err
		}
		if configuration.Build.Command != "" {
			return invalidRunConfiguration("build.command", "is not allowed for ant")
		}
	case "javac":
		if configuration.Build.Target != "" || configuration.Build.Command != "" {
			return invalidRunConfiguration("build", "javac must not define target or command")
		}
	case "custom":
		if err := validateNonBlank("build.command", configuration.Build.Command, 4096); err != nil {
			return err
		}
		if configuration.Build.Target != "" {
			return invalidRunConfiguration("build.target", "is not allowed for custom")
		}
	default:
		return invalidRunConfiguration("build.type", "must be ant, javac or custom")
	}

	if err := validateStableID("server.id", configuration.Server.ID); err != nil {
		return err
	}
	if configuration.Server.HTTPPort < 1024 || configuration.Server.HTTPPort > 65535 {
		return invalidRunConfiguration("server.httpPort", "must be between 1024 and 65535")
	}
	if configuration.Server.DebugPort < 1024 || configuration.Server.DebugPort > 65535 {
		return invalidRunConfiguration("server.debugPort", "must be between 1024 and 65535")
	}
	if configuration.Server.HTTPPort == configuration.Server.DebugPort {
		return invalidRunConfiguration("server.debugPort", "must differ from httpPort")
	}
	if !contextPathPattern.MatchString(configuration.Server.ContextPath) {
		return invalidRunConfiguration("server.contextPath", "has an invalid context path")
	}

	if configuration.Deploy.Mode != "exploded" && configuration.Deploy.Mode != "war" {
		return invalidRunConfiguration("deploy.mode", "must be exploded or war")
	}
	if err := validateArtifact(configuration.Deploy.Artifact); err != nil {
		return err
	}
	if configuration.Env == nil {
		return invalidRunConfiguration("env", "is required")
	}
	if len(configuration.Env) > 128 {
		return invalidRunConfiguration("env", "must contain at most 128 entries")
	}
	seenEnvironmentNames := make(map[string]struct{}, len(configuration.Env))
	for name, value := range configuration.Env {
		if !environmentNamePattern.MatchString(name) {
			return invalidRunConfiguration("env."+name, "has an invalid name")
		}
		if utf8.RuneCountInString(value) > 8192 {
			return invalidRunConfiguration("env."+name, "must be at most 8192 characters")
		}
		folded := strings.ToUpper(name)
		if _, exists := seenEnvironmentNames[folded]; exists {
			return invalidRunConfiguration("env", "contains names that conflict by case")
		}
		seenEnvironmentNames[folded] = struct{}{}
		if sensitiveEnvironmentName.MatchString(name) && !environmentReferencePattern.MatchString(value) {
			return invalidRunConfiguration("env."+name, "sensitive values must use a ${env:HOST_NAME} reference")
		}
	}
	if configuration.VMOptions == nil {
		return invalidRunConfiguration("vmOptions", "is required")
	}
	if len(configuration.VMOptions) > 128 {
		return invalidRunConfiguration("vmOptions", "must contain at most 128 values")
	}
	for index, option := range configuration.VMOptions {
		if err := validateNonBlank(fmt.Sprintf("vmOptions[%d]", index), option, 2048); err != nil {
			return err
		}
	}
	if configuration.BeforeLaunchTasks == nil {
		return invalidRunConfiguration("beforeLaunchTasks", "is required")
	}
	if len(configuration.BeforeLaunchTasks) > 2 {
		return invalidRunConfiguration("beforeLaunchTasks", "must contain at most two tasks")
	}
	taskIndex := map[string]int{}
	for index, task := range configuration.BeforeLaunchTasks {
		if task != "build" && task != "deploy" {
			return invalidRunConfiguration(fmt.Sprintf("beforeLaunchTasks[%d]", index), "must be build or deploy")
		}
		if _, exists := taskIndex[task]; exists {
			return invalidRunConfiguration("beforeLaunchTasks", "must contain unique tasks")
		}
		taskIndex[task] = index
	}
	if buildIndex, hasBuild := taskIndex["build"]; hasBuild {
		if deployIndex, hasDeploy := taskIndex["deploy"]; hasDeploy && deployIndex < buildIndex {
			return invalidRunConfiguration("beforeLaunchTasks", "deploy must not run before build")
		}
	}
	return nil
}

func (document RunConfigurationDocument) Validate() error {
	if document.Version != RunConfigurationVersion {
		return invalidRunConfiguration("version", "must equal 1")
	}
	if document.Configurations == nil {
		return invalidRunConfiguration("configurations", "is required")
	}
	if len(document.Configurations) > 100 {
		return invalidRunConfiguration("configurations", "must contain at most 100 entries")
	}
	ids := make(map[string]struct{}, len(document.Configurations))
	names := make(map[string]struct{}, len(document.Configurations))
	for index, configuration := range document.Configurations {
		if err := configuration.Validate(); err != nil {
			return fmt.Errorf("configurations[%d]: %w", index, err)
		}
		if _, exists := ids[configuration.ID]; exists {
			return invalidRunConfiguration(fmt.Sprintf("configurations[%d].id", index), "must be unique")
		}
		ids[configuration.ID] = struct{}{}
		foldedName := strings.ToLower(configuration.Name)
		if _, exists := names[foldedName]; exists {
			return invalidRunConfiguration(fmt.Sprintf("configurations[%d].name", index), "must be unique ignoring case")
		}
		names[foldedName] = struct{}{}
	}
	if len(document.Configurations) == 0 {
		if document.SelectedConfigurationID != nil {
			return invalidRunConfiguration("selectedConfigurationId", "must be null when configurations is empty")
		}
	} else {
		if document.SelectedConfigurationID == nil {
			return invalidRunConfiguration("selectedConfigurationId", "must reference an existing configuration")
		}
		if _, exists := ids[*document.SelectedConfigurationID]; !exists {
			return invalidRunConfiguration("selectedConfigurationId", "must reference an existing configuration")
		}
	}
	return nil
}

func requireFields(object map[string]json.RawMessage, path string, fields ...string) error {
	for _, field := range fields {
		if _, exists := object[field]; !exists {
			return invalidRunConfiguration(path+field, "is required")
		}
	}
	return nil
}

func validateRequiredJSONFields(data []byte) error {
	var top map[string]json.RawMessage
	if err := json.Unmarshal(data, &top); err != nil {
		return err
	}
	if err := requireFields(top, "", "version", "configurations", "selectedConfigurationId"); err != nil {
		return err
	}
	var configurations []json.RawMessage
	if err := json.Unmarshal(top["configurations"], &configurations); err != nil {
		return err
	}
	for index, raw := range configurations {
		var configuration map[string]json.RawMessage
		if err := json.Unmarshal(raw, &configuration); err != nil {
			return err
		}
		prefix := fmt.Sprintf("configurations[%d].", index)
		if err := requireFields(configuration, prefix, "id", "name", "type", "projectId", "mode", "suspend", "jdkRef", "build", "server", "deploy", "env", "vmOptions", "beforeLaunchTasks"); err != nil {
			return err
		}
		for name, fields := range map[string][]string{
			"build": {"type", "clean"}, "server": {"id", "httpPort", "debugPort", "contextPath"}, "deploy": {"mode", "artifact"},
		} {
			var nested map[string]json.RawMessage
			if err := json.Unmarshal(configuration[name], &nested); err != nil {
				return err
			}
			if err := requireFields(nested, prefix+name+".", fields...); err != nil {
				return err
			}
		}
	}
	return nil
}

func decodeStrict(data []byte, target any) error {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(target); err != nil {
		return err
	}
	var extra any
	if err := decoder.Decode(&extra); err != io.EOF {
		return fmt.Errorf("multiple JSON values are not allowed")
	}
	return nil
}

func DecodeRunConfigurationDocument(data []byte) (RunConfigurationDocument, error) {
	var document RunConfigurationDocument
	if err := validateRequiredJSONFields(data); err != nil {
		return document, fmt.Errorf("%w: %v", ErrInvalidRunConfiguration, err)
	}
	if err := decodeStrict(data, &document); err != nil {
		return document, fmt.Errorf("%w: %v", ErrInvalidRunConfiguration, err)
	}
	if err := document.Validate(); err != nil {
		return document, err
	}
	return document, nil
}

func DecodeTomcatRunConfiguration(data []byte) (TomcatRunConfiguration, error) {
	wrapped := []byte(fmt.Sprintf(`{"version":1,"configurations":[%s],"selectedConfigurationId":null}`, data))
	// Validate required nested fields using the document decoder, then repair the
	// selection solely for semantic validation of this one configuration.
	if err := validateRequiredJSONFields(wrapped); err != nil {
		return TomcatRunConfiguration{}, fmt.Errorf("%w: %v", ErrInvalidRunConfiguration, err)
	}
	var configuration TomcatRunConfiguration
	if err := decodeStrict(data, &configuration); err != nil {
		return configuration, fmt.Errorf("%w: %v", ErrInvalidRunConfiguration, err)
	}
	if err := configuration.Validate(); err != nil {
		return configuration, err
	}
	return configuration, nil
}

File: packages/config-schema/src/run-configuration.ts
import Ajv2020, { type ErrorObject, type ValidateFunction } from 'ajv/dist/2020';
import {
  type RunConfigurationDocument,
  type TomcatRunConfiguration,
} from '@kairo/protocol';

export const RUN_CONFIGURATION_FILE = '.legacyflow/run-configurations.json' as const;
/** Must remain aligned with @kairo/protocol.RUN_CONFIGURATION_VERSION. */
export const RUN_CONFIGURATION_SCHEMA_VERSION = 1 as const;

const nonBlank = '^\\S(?:.*\\S)?$';
const stableId = '^[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$';
const relativePath = '^(?![A-Za-z]:[\\\\/])(?![\\\\/])\\S(?:.*\\S)?$';

const buildBase = {
  type: 'object',
  additionalProperties: false,
  required: ['type', 'clean'],
  properties: {
    type: { type: 'string' },
    clean: { type: 'boolean' },
  },
} as const;

export const runConfigurationJsonSchema = {
  $schema: 'https://json-schema.org/draft/2020-12/schema',
  $id: 'https://kairo.local/schemas/run-configurations/v1.json',
  title: 'Kairo Run Configuration Document',
  type: 'object',
  additionalProperties: false,
  required: ['version', 'configurations', 'selectedConfigurationId'],
  properties: {
    version: { const: RUN_CONFIGURATION_SCHEMA_VERSION },
    configurations: {
      type: 'array',
      maxItems: 100,
      items: { $ref: '#/$defs/tomcatRunConfiguration' },
    },
    selectedConfigurationId: {
      oneOf: [{ type: 'string', pattern: stableId }, { type: 'null' }],
    },
  },
  $defs: {
    build: {
      oneOf: [
        {
          ...buildBase,
          required: ['type', 'target', 'clean'],
          properties: {
            ...buildBase.properties,
            type: { const: 'ant' },
            target: { type: 'string', minLength: 1, maxLength: 200, pattern: nonBlank },
          },
        },
        {
          ...buildBase,
          properties: { ...buildBase.properties, type: { const: 'javac' } },
        },
        {
          ...buildBase,
          required: ['type', 'command', 'clean'],
          properties: {
            ...buildBase.properties,
            type: { const: 'custom' },
            command: { type: 'string', minLength: 1, maxLength: 4096, pattern: nonBlank },
          },
        },
      ],
    },
    tomcatRunConfiguration: {
      type: 'object',
      additionalProperties: false,
      required: [
        'id', 'name', 'type', 'projectId', 'mode', 'suspend', 'jdkRef',
        'build', 'server', 'deploy', 'env', 'vmOptions', 'beforeLaunchTasks',
      ],
      properties: {
        id: { type: 'string', pattern: stableId },
        name: { type: 'string', minLength: 1, maxLength: 200, pattern: nonBlank },
        type: { const: 'tomcat6' },
        projectId: { type: 'string', pattern: stableId },
        mode: { enum: ['run', 'debug'] },
        suspend: { type: 'boolean' },
        jdkRef: { type: 'string', pattern: stableId },
        build: { $ref: '#/$defs/build' },
        server: {
          type: 'object',
          additionalProperties: false,
          required: ['id', 'httpPort', 'debugPort', 'contextPath'],
          properties: {
            id: { type: 'string', pattern: stableId },
            httpPort: { type: 'integer', minimum: 1024, maximum: 65535 },
            debugPort: { type: 'integer', minimum: 1024, maximum: 65535 },
            contextPath: { type: 'string', pattern: '^/[A-Za-z0-9._-]*$' },
          },
        },
        deploy: {
          type: 'object',
          additionalProperties: false,
          required: ['mode', 'artifact'],
          properties: {
            mode: { enum: ['exploded', 'war'] },
            artifact: { type: 'string', minLength: 1, maxLength: 1024, pattern: relativePath },
          },
        },
        env: {
          type: 'object',
          maxProperties: 128,
          propertyNames: { pattern: '^[A-Za-z_][A-Za-z0-9_]*$' },
          additionalProperties: { type: 'string', maxLength: 8192 },
        },
        vmOptions: {
          type: 'array',
          maxItems: 128,
          items: { type: 'string', minLength: 1, maxLength: 2048, pattern: nonBlank },
        },
        beforeLaunchTasks: {
          type: 'array',
          maxItems: 2,
          uniqueItems: true,
          items: { enum: ['build', 'deploy'] },
        },
      },
      allOf: [
        {
          if: { properties: { mode: { const: 'run' } }, required: ['mode'] },
          then: { properties: { suspend: { const: false } } },
        },
      ],
    },
  },
} as const;

export interface RunConfigurationValidationIssue {
  path: string;
  message: string;
}

export type RunConfigurationValidationResult =
  | { valid: true; value: RunConfigurationDocument; issues: [] }
  | { valid: false; issues: RunConfigurationValidationIssue[] };

export class RunConfigurationValidationError extends Error {
  constructor(public readonly issues: RunConfigurationValidationIssue[]) {
    super(`Invalid run configuration: ${issues.map(issue => `${issue.path} ${issue.message}`).join('; ')}`);
    this.name = 'RunConfigurationValidationError';
  }
}

const ajv = new Ajv2020({ allErrors: true, strict: true });
const validateSchema: ValidateFunction<RunConfigurationDocument> = ajv.compile(runConfigurationJsonSchema);

function schemaIssues(errors: ErrorObject[] | null | undefined): RunConfigurationValidationIssue[] {
  return (errors || []).map(error => ({
    path: error.instancePath || '/',
    message: error.message || error.keyword,
  }));
}

function semanticIssues(document: RunConfigurationDocument): RunConfigurationValidationIssue[] {
  const issues: RunConfigurationValidationIssue[] = [];
  const ids = new Set<string>();
  const names = new Set<string>();
  for (const [index, configuration] of document.configurations.entries()) {
    const base = `/configurations/${index}`;
    if (ids.has(configuration.id)) issues.push({ path: `${base}/id`, message: 'must be unique' });
    ids.add(configuration.id);
    const foldedName = configuration.name.toLocaleLowerCase('en-US');
    if (names.has(foldedName)) issues.push({ path: `${base}/name`, message: 'must be unique ignoring case' });
    names.add(foldedName);
    if (configuration.server.httpPort === configuration.server.debugPort) {
      issues.push({ path: `${base}/server/debugPort`, message: 'must differ from httpPort' });
    }
    const artifactSegments = configuration.deploy.artifact.split(/[\\/]+/);
    if (configuration.deploy.artifact.includes('\\')) {
      issues.push({ path: `${base}/deploy/artifact`, message: 'must use forward slashes' });
    }
    if (configuration.deploy.artifact.includes(':')) {
      issues.push({ path: `${base}/deploy/artifact`, message: 'must not contain a colon' });
    }
    if (artifactSegments.includes('..')) {
      issues.push({ path: `${base}/deploy/artifact`, message: 'must not escape the project root with .. segments' });
    }
    const buildIndex = configuration.beforeLaunchTasks.indexOf('build');
    const deployIndex = configuration.beforeLaunchTasks.indexOf('deploy');
    if (buildIndex >= 0 && deployIndex >= 0 && deployIndex < buildIndex) {
      issues.push({ path: `${base}/beforeLaunchTasks`, message: 'deploy must not run before build' });
    }
    const envNames = new Set<string>();
    for (const [name, value] of Object.entries(configuration.env)) {
      const folded = name.toLocaleUpperCase('en-US');
      if (envNames.has(folded)) {
        issues.push({ path: `${base}/env`, message: `environment variable ${name} conflicts by case` });
      }
      envNames.add(folded);
      const sensitive = /(PASSWORD|PASSWD|SECRET|TOKEN|API_KEY|PRIVATE_KEY)/i.test(name);
      if (sensitive && !/^\$\{env:[A-Za-z_][A-Za-z0-9_]*\}$/.test(value)) {
        issues.push({ path: `${base}/env/${name}`, message: 'sensitive values must use a ${env:HOST_NAME} reference' });
      }
    }
  }
  if (document.configurations.length === 0 && document.selectedConfigurationId !== null) {
    issues.push({ path: '/selectedConfigurationId', message: 'must be null when configurations is empty' });
  } else if (document.configurations.length > 0 && !ids.has(document.selectedConfigurationId || '')) {
    issues.push({ path: '/selectedConfigurationId', message: 'must reference an existing configuration' });
  }
  return issues;
}

export function validateRunConfigurationDocument(input: unknown): RunConfigurationValidationResult {
  if (!validateSchema(input)) return { valid: false, issues: schemaIssues(validateSchema.errors) };
  const issues = semanticIssues(input);
  return issues.length === 0 ? { valid: true, value: input, issues: [] } : { valid: false, issues };
}

export function parseRunConfigurationDocument(text: string): RunConfigurationDocument {
  let input: unknown;
  try {
    input = JSON.parse(text) as unknown;
  } catch (error) {
    throw new RunConfigurationValidationError([{
      path: '/',
      message: `must be valid JSON (${error instanceof Error ? error.message : String(error)})`,
    }]);
  }
  const result = validateRunConfigurationDocument(input);
  if (!result.valid) throw new RunConfigurationValidationError(result.issues);
  return result.value;
}

function canonicalConfiguration(configuration: TomcatRunConfiguration): TomcatRunConfiguration {
  return {
    id: configuration.id,
    name: configuration.name,
    type: configuration.type,
    projectId: configuration.projectId,
    mode: configuration.mode,
    suspend: configuration.suspend,
    jdkRef: configuration.jdkRef,
    build: { ...configuration.build },
    server: { ...configuration.server },
    deploy: { ...configuration.deploy },
    env: Object.fromEntries(Object.entries(configuration.env).sort(([left], [right]) => left.localeCompare(right))),
    vmOptions: [...configuration.vmOptions],
    beforeLaunchTasks: [...configuration.beforeLaunchTasks],
  };
}

export function serializeRunConfigurationDocument(document: RunConfigurationDocument): string {
  const result = validateRunConfigurationDocument(document);
  if (!result.valid) throw new RunConfigurationValidationError(result.issues);
  const canonical: RunConfigurationDocument = {
    version: RUN_CONFIGURATION_SCHEMA_VERSION,
    configurations: result.value.configurations.map(canonicalConfiguration),
    selectedConfigurationId: result.value.selectedConfigurationId,
  };
  return `${JSON.stringify(canonical, null, 2)}\n`;
}
